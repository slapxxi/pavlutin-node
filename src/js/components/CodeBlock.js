import React from 'react';
import Prism from 'prismjs';
import 'prismjs/components/prism-jsx.min';

function CodeBlock(props) {
  const { language, literal } = props;
  const clsName = `language-${language || 'text'}`;
  let parsingLanguage;
  let markup;
  if (language === 'js') {
    parsingLanguage = Prism.languages.javascript;
  }
  if (language === 'css') {
    parsingLanguage = Prism.languages.css;
  }
  if (language === 'html') {
    parsingLanguage = Prism.languages.html;
  }
  if (language === 'jsx') {
    parsingLanguage = Prism.languages.jsx;
  }
  if (parsingLanguage) {
    markup = { __html: Prism.highlight(literal, parsingLanguage) };
    return (
      <div className="codeblock">
        <pre className={clsName}>
          <code
            className={clsName}
            dangerouslySetInnerHTML={markup}
          />
        </pre>
        <div className="codeblock__nav">
          { language }
        </div>
      </div>
    );
  }
  return (
    <div className="codeblock">
      <pre className={clsName}>
        <code className={clsName}>
          { literal }
        </code>
      </pre>
    </div>
  );
}

export default CodeBlock;
