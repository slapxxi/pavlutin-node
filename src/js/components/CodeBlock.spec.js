import React from 'react';
import { shallow } from 'enzyme';
import CodeBlock from './CodeBlock';
import { singleLine } from '../../../test/test-helper';

const jsSnippet = "const test = 'test';";
const cssSnippet = singleLine`
  h1 {
    color: red;
  }
`;

it('renders', () => {
  const wrapper = shallow(<CodeBlock />);
  expect(wrapper.length).toBe(1);
});

it('sets class matching language prop', () => {
  const wrapper = shallow(<CodeBlock language="js" literal="" />);
  expect(wrapper.find('pre.language-js').length).toBe(1);
});

it('sets class to "text" if missing language', () => {
  const wrapper = shallow(<CodeBlock literal="" />);
  expect(wrapper.find('pre.language-text').length).toBe(1);
});

it('converts js literal to language tokens', () => {
  const wrapper = shallow(<CodeBlock language="js" literal={jsSnippet} />);
  const expected = singleLine`
    <pre class="language-js">
      <code class="language-js">
        <span class="token keyword">const</span> test <
        span class="token operator">=</span> <
        span class="token string">'test'</span>
        <span class="token punctuation">;</span>
      </code>
    </pre>
  `;
  expect(wrapper.html()).toContain(expected);
});

it('converts css literal to language tokens', () => {
  const wrapper = shallow(<CodeBlock language="css" literal={cssSnippet} />);
  const expected = singleLine`
    <pre class="language-css">
      <code class="language-css">
        <span class="token selector">h1</span> <
        span class="token punctuation">{</span>
        <span class="token property">color</span>
        <span class="token punctuation">:</span> red
        <span class="token punctuation">;</span>
        <span class="token punctuation">}</span>
      </code>
    </pre>
  `;
  expect(wrapper.html()).toContain(expected);
});
