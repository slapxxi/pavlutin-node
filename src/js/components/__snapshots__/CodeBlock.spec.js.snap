// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="codeblock"
>
  <pre
    className="language-text"
  >
    <code
      className="language-text"
    />
  </pre>
</div>
`;
